mobile_verifier/src/rewarder.rs

```diff
@@ -0,0 +1,249 @@
+use crate::{
+    heartbeats::HeartbeatReward,
+    ingest,
+    reward_shares::{PocShares, TransferRewards},
+    speedtests::SpeedtestAverages,
+};
+use anyhow::bail;
+use chrono::{DateTime, Duration, TimeZone, Utc};
+use db_store::meta;
+use file_store::{file_sink::FileSinkClient, traits::TimestampEncode, FileStore};
+use helium_proto::RewardManifest;
+use price::PriceTracker;
+use reward_scheduler::Scheduler;
+use rust_decimal::{prelude::ToPrimitive, Decimal};
+use rust_decimal_macros::dec;
+use sqlx::{PgExecutor, Pool, Postgres};
+use std::ops::Range;
+use tokio::time::sleep;
+
+const REWARDS_NOT_CURRENT_DELAY_PERIOD: i64 = 5;
+
+pub struct Rewarder {
+    pool: Pool<Postgres>,
+    reward_period_duration: Duration,
+    reward_offset: Duration,
+    mobile_rewards: FileSinkClient,
+    reward_manifests: FileSinkClient,
+    price_tracker: PriceTracker,
+    data_transfer_ingest: FileStore,
+}
+
+impl Rewarder {
+    pub fn new(
+        pool: Pool<Postgres>,
+        reward_period_duration: Duration,
+        reward_offset: Duration,
+        mobile_rewards: FileSinkClient,
+        reward_manifests: FileSinkClient,
+        price_tracker: PriceTracker,
+        data_transfer_ingest: FileStore,
+    ) -> Self {
+        Self {
+            pool,
+            reward_period_duration,
+            reward_offset,
+            mobile_rewards,
+            reward_manifests,
+            price_tracker,
+            data_transfer_ingest,
+        }
+    }
+
+    pub async fn run(self, shutdown: triggered::Listener) -> anyhow::Result<()> {
+        loop {
+            let last_rewarded_end_time = self.last_rewarded_end_time().await?;
+            let next_rewarded_end_time = self.next_rewarded_end_time().await?;
+            let scheduler = Scheduler::new(
+                self.reward_period_duration,
+                last_rewarded_end_time,
+                next_rewarded_end_time,
+                self.reward_offset,
+            );
+            let now = Utc::now();
+            let sleep_duration = if scheduler.should_reward(now) {
+                if self.is_data_current(&scheduler.reward_period).await? {
+                    self.reward(&scheduler).await?;
+                    continue;
+                } else {
+                    tracing::info!(
```

Comment:
do we log at `info` level by default? if not this should probably be a warning

Reply:
yeah, I don't really think this is a warning. Although maybe it is? I don't expect this to happen too often, if at all